main/src/ca/uwaterloo/flix/language/phase/typer/SubstitutionTree.scala

```diff
@@ -33,8 +33,18 @@ case class SubstitutionTree(root: Substitution, branches: Map[Symbol.KindedTypeV
     * Applies this substitution tree to the given type constraint.
     */
   def apply(constr: TypeConstraint2): TypeConstraint2 = constr match {
-    case TypeConstraint2.Equality(tpe1, tpe2, loc) => TypeConstraint2.Equality(root(tpe1), root(tpe2), loc)
-    case TypeConstraint2.Trait(sym, tpe, loc) => TypeConstraint2.Trait(sym, root(tpe), loc)
+    case TypeConstraint2.Equality(tpe1, tpe2, loc) =>
+      // Performance: Reuse this, if possible.
+      val t1 = root(tpe1)
+      val t2 = root(tpe2)
+      if ((t1 eq tpe1) && (t2 eq tpe2))
```

Comment:
We have this pattern a lot at this point, and it is a potential place for typos (different names for the same conceptual thing (e.g. `t1`, `tpe1`). Could we create a Util function for it?

Reply:
I had the same thought, but I don't see how. It really feels like belongs in the compiler somehow.